shachain/store.go

```diff
@@ -0,0 +1,187 @@
+package shachain
+
+import (
+	"bytes"
+	"encoding/binary"
+	"github.com/go-errors/errors"
+	"github.com/roasbeef/btcd/wire"
+)
+
+// SecretStore is an interface which serves as an abstraction over data
+// structure responsible for effecient storing and restoring of hash secrets
+// by given indexes.
+//
+// Description: The Lightning Network wants a chain of (say 1 million)
+// unguessable 256 bit values; we generate them and send them one at a time
+// to a remote node.  We don't want the remote node to have to store all the
+// values, so it's better if they can derive them once they see them.
+type SecretStore interface {
+	// Restore function is ised to restore the previous secret by its index.
+	Restore(uint64) (*wire.ShaHash, error)
+
+	// Store is used to store the given sha hash in efficient maner.
+	Store(*wire.ShaHash) error
+
+	// ToBytes convert store to the binary representation.
+	ToBytes() ([]byte, error)
+}
+
+// ShaChainStore implementation of SecretStore. This version of shachain store
+// sligtly changed in terms of method naming. Initial concept might be found
+// here:
+// https://github.com/rustyrussell/ccan/blob/master/ccan/crypto/shachain/design.txt
+type ShaChainStore struct {
```

Comment:
In a similar vein to my comment above, I thin kthis should be renamed to `Receiver`. When referenced out of the package, the full import string will not have any redundancy: `shachain.Receiver`